numpy/core/src/multiarray/multiarraymodule.c

```diff
@@ -1271,6 +1271,13 @@ PyArray_Correlate2(PyObject *op1, PyObject *op2, int mode)
     typenum = PyArray_ObjectType(op2, typenum);
 
     typec = PyArray_DescrFromType(typenum);
+    if (typec == NULL) {
+        if (!PyErr_Occurred()) {
+            PyErr_SetString(PyExc_TypeError,
+                            "Cannot find a common data type.");
+        }
```

Comment:
This should never happen. If there is a path through `PyArray_DescrFromType` that returns NULL and does not set the error, that should be fixed.

Reply:
Yeah, it is strange, but the `PyArray_DescrFromType` has a specific comment there that an error is not set for backward compatibility in the C-API (I think I will disagree and say that we can change that part, but it might be better with a DeprecationWarning).

In any case, we should not do this here, `PyArray_DescrFromType` can (unlikely) fail, and the NULL return if typecode is NULL is good. The no error set path can only happen if `NPY_NOTYPE`  is passed in, and that must be checked above already after each `PyArray_ObjectType` call. If `NPY_NOTYPE` is returned, then an error should be set already.
(That code is a bit of a rabbit hole, so I won't rule out that there is some difficulty, but I do not expect it).

I saw that this function also has "nice" overloading in that you can write `PyArray_DescrFromType('?')` and it will work with nothing more but a compiler warning maybe ;))

Reply:
@seberg it is not clear to me what the way forward here is. Could you show what you mean in code?

Reply:
If @Soniyanayak51 may want to revive it. What I meant was to move the error check further up.

But looking at the code again, what I think should happen is that the `PyArray_ObjectType(...)` calls should both be replaced by `PyArray_DTypeFromObject` calls (which returns -1 on error). `PyArray_ObjectType` should be deprecated, but that is a different issue.